src/vsr/grid_scrubber.zig

```diff
@@ -0,0 +1,481 @@
+//!   data is unavailable.
+//! - "Latent" refers to: the error is not discoverable until the sector is actually read.
+//! - "An Analysis of Latent Sector Errors in Disk Drives" (2007) found that >60% of latent sector
+//!   errors were discovered by a scrubber that cycles every 2 weeks.
+//!
+//! Finding and repairing errors proactively minimizes the risk of cluster data loss due to multiple
+//! intersecting faults (analogous to a "double-fault") – a scenario where we fail to read a block,
+//! and try to repair the block from another replica, only to discover that the copy of the block on
+//! the remote replica's disk is *also* faulty.
+//!
+//! TODO Start replicas scrubbing from distinct/random offsets in the tour to farther minimize risk
+//! of cluster data loss. (Right now replicas will often have identical scrubbing schedules.)
+const std = @import("std");
+const assert = std.debug.assert;
+const maybe = stdx.maybe;
+const log = std.log.scoped(.grid_scrubber);
+
+const stdx = @import("../stdx.zig");
+const vsr = @import("../vsr.zig");
+const schema = @import("../lsm/schema.zig");
+const FIFO = @import("../fifo.zig").FIFO;
+const RingBuffer = @import("../ring_buffer.zig").RingBuffer;
+
+const allocate_block = @import("./grid.zig").allocate_block;
+const GridType = @import("./grid.zig").GridType;
+const BlockPtr = @import("./grid.zig").BlockPtr;
+const ForestTableIteratorType = @import("../lsm/forest_table_iterator.zig").ForestTableIteratorType;
+const snapshot_from_op = @import("../lsm/manifest.zig").snapshot_from_op;
+const TestStorage = @import("../testing/storage.zig").Storage;
+
+pub fn GridScrubberType(comptime Forest: type) type {
+    return struct {
+        const GridScrubber = @This();
+        const Grid = GridType(Forest.Storage);
+        const ForestTableIterator = ForestTableIteratorType(Forest);
+        const SuperBlock = vsr.SuperBlockType(Forest.Storage);
+        const ManifestBlockIterator = ManifestBlockIteratorType(Forest.ManifestLog);
+        const CheckpointTrailer = vsr.CheckpointTrailerType(Forest.Storage);
+
+        pub const BlockId = struct {
+            block_checksum: u128,
+            block_address: u64,
+            block_type: schema.BlockType,
+        };
+
+        pub const Read = struct {
+            scrubber: *GridScrubber,
+            read: Grid.Read = undefined,
+            block_type: ?schema.BlockType,
+
+            /// Set to true before the read completes if the scrub result should be ignored:
+            /// - The scrub succeeded – no need for repair.
+            /// - The block was freed by a checkpoint in the time that the read was in progress.
+            ///   (At checkpoint, the FreeSet frees blocks released during the preceding checkpoint.
+            ///   We can scrub released blocks, but not free blocks. Setting this flag ensures that
+            ///   GridScrubber doesn't require a read-barrier at checkpoint.)
+            ignore: bool,
```

Comment:
To make this a bit more self-explanatory and greppable, might use a `status: enum { ok, canceled, overwritten,, ... }` here.